### PR TITLE
neuron: coreneuron only supports nmodl

### PR DESCRIPTION
**Pull request by alkino**
* neuron: in https://github.com/neuronsimulator/nrn/pull/2247 CoreNEURON dropped support for mod2c
  * This means that `neuron@develop +coreneuron` always uses nmodl
  * Variants have been made conditional so that meaningless things like `neuron@develop +coreneuron~nmodl` are not allowed

### CODE REVIEW

File: bluebrain/repo-patches/packages/neuron/package.py [hunk 1]
```diff
@@ -93,20 +93,21 @@ class Neuron(CMakePackage):
...
+    variant("openmp", default=False, description="Enable OpenMP support", when="@9:")
     variant("report", default=True, description="Enable SONATA and binary reports")
     variant("shared", default=True, description="Build shared library")
-    variant("nmodl", default=True, description="Use NMODL instead of MOD2C")
+    variant("nmodl", default=True, description="Use NMODL instead of MOD2C", when="@9:9.0.a6 +coreneuron")
     variant(
         "codegenopt",
         default=False,
         description="Use NMODL with codedgen ionvar copies optimizations",
+        when="+nmodl",
```
Comment (olupton) [inline bluebrain/repo-patches/packages/neuron/package.py]: `+codegenopt` also makes sense for the latest/develop/mod2c-free version, which doesn't have `+nmodl`.

I think `@9:9.0.a6 +coreneuron+nmodl` and `@develop +coreneuron` (I'm assuming `.a6` still supports `mod2c`) should both allow `+codegenopt`.

File: bluebrain/repo-patches/packages/neuron/package.py [hunk 2]
```diff
@@ -93,20 +93,21 @@ class Neuron(CMakePackage):
...
     )
-    variant("sympy", default=False, description="Use NMODL with SymPy to solve ODEs")
-    variant("sympyopt", default=False, description="Use NMODL with SymPy Optimizations")
+    variant("sympy", default=False, description="Use NMODL with SymPy to solve ODEs", when="@9: +nmodl")
```
Comment (olupton) [inline bluebrain/repo-patches/packages/neuron/package.py]: Same comment as `codegenopt`, I think this makes sense in all the same cases.

File: bluebrain/repo-patches/packages/neuron/package.py [hunk 3]
```diff
@@ -156,29 +157,14 @@ class Neuron(CMakePackage):
     depends_on("python", type=("build", "run"))
     depends_on("boost", when="@8.99:+tests+coreneuron")
     depends_on("cuda", when="@8.99:+gpu")
-    depends_on("flex@2.6:", type="build", when="+nmodl")
-    depends_on("nmodl@0.4.0:", when="@8.99:+nmodl")
+    depends_on("nmodl@0.4.0:", when="@9:+nmodl")
```
Comment (olupton) [inline bluebrain/repo-patches/packages/neuron/package.py]: We need to depend on `nmodl` even in the `develop` version that doesn't have an `+nmodl` variant.

**Review by olupton — COMMENTED** [review]
I think there are still some wrinkles to iron out.
Alternatively, if it would simplify things, we could leave the `nmodl` variant in and just emit an error if someone tries to build `@develop ~nmodl` etc.

File: bluebrain/repo-patches/packages/neuron/package.py (hunk 1 above)
Comment (olupton) [reply bluebrain/repo-patches/packages/neuron/package.py]: If the result of the `when` clause on the `nmodl` `variant` is that for `@develop`, `+nmodl` is always true, then we can mostly ignore the above comment.

Only "mostly" in the sense that I think `neuron@8.2.2 +codegenopt` would currently be allowed?

**Review by alexsavulescu — DISMISSED** [review]
LGTM

File: bluebrain/deployment/gitlab-ci.yml [hunk 4]
```diff
@@ -393,6 +393,7 @@ check_neuron_ci:
     - if: $TRIGGER_CHILD_NEURON_PIPELINE
   trigger:
     project: hpc/cellular/nrn
+    branch: cornu/remove_mod2c
```
Comment (alexsavulescu) [inline bluebrain/deployment/gitlab-ci.yml]: reminder to revert this

**Review by olupton — APPROVED** [review]

**Pull request merged**